### PR TITLE
Interval Utils (round 2)

### LINKED ISSUE #464: Interval API
At the moment, we have Plutarch versions of `Interval`, `LowerBound`, `UpperBound`, etc. but none of the useful functions available in `plutus-ledger-api`, like `intersects`, `contains` and others. It would be nice to have these (whether in `plutarch` or `plutarch-extras`), since any project that needs to validate time-sensitive operations would have to implement them.

### PR DESCRIPTION
**Pull request by MangoIV**
- reviving #474 
- closes #464

### CODE REVIEW

**Review by L-as — APPROVED** [review]
I sure hope this is correct.

**Review by t1lde — COMMENTED** [review]
quick review, didn't spot anything off. 

- Could it be worth testing equivalence with the PlutusLedgerApi versions?
- Since intervals for a lattice with join = hull, bottom = never, meet = intersection, top = always, you could test the lattice laws - as in the  [purescript](https://github.com/Risto-Stevcev/purescript-lattice/blob/v0.3.0/src/Data/Lattice/Verify.purs#L9-L9) and [haskell](https://hackage.haskell.org/package/lattices-2.0.3/docs/Algebra-Lattice.html#g:1) Lattice packages

**Review by TotallyNotChase — APPROVED** [review]
it's possible that some functions could be optimized further but logic looks good

File: plutarch-test/plutarch-extra/Plutarch/Extra/IntervalSpec.hs [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+          i4 :: Term s (PInterval PInteger)
+          i4 = mkInterval 4 4
+          i5 :: Term s (PInterval PInteger)
+          i5 = mkInterval 3 4
+      pgoldenSpec $ do
+        "constants" @\ do
+          "always" @| palways @PInteger @-> psucceeds
+          "never" @| pnever @PInteger @-> psucceeds
+          "always" @| pcontains # palways @PInteger # i1 @-> passert
+          "never" @| pcontains # pnever @PInteger # i1 @-> passertNot
```
[the remark this answers is not in the document]
Comment (MangoIV) [reply plutarch-test/plutarch-extra/Plutarch/Extra/IntervalSpec.hs:47]: good catch, that's correct!

**Pull request merged**